lighthouse-logger/index.js

```diff
@@ -118,33 +174,164 @@ class Log {
     return level_ === 'verbose';
   }
 
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
   static time({msg, id, args = []}, level = 'log') {
     marky.mark(id);
     Log[level]('status', msg, ...args);
   }
 
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
   static timeEnd({msg, id, args = []}, level = 'verbose') {
     Log[level]('statusEnd', msg, ...args);
     marky.stop(id);
   }
 
+  /* eslint-disable no-invalid-this */
+  /**
+   * Decorates a function, calling time/timeEnd before/after calling the original function.
+   * @template T, R
+   * @template {*[]} Args
+   * @param {(this: T, ...args: Args) => R} originalFn
+   * @param {TimeDecorateOpts<T, Args>} opts
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: *, ...args: *[]) => string}
+     */
+    const computeMsg = (_this, args) => {
```

Comment:
could this fn be as simple as:

```js
  const computeMsg = (_this, args) => {
    if (typeof opts.msg === 'string') return opts.msg;
    if (typeof opts.msg === 'function') return opts.msg.apply(_this, args);
    throw new Error('expected msg');
  };
```

and same with computeId()

checking that the first two cases don't return something falsy seems.. unnecessary to me, but w/e

Reply:
oh ya for sure

this only made sense before I figured out how to strongly type all this. had to verify this at runtime. but now TSC does the heavy lifting here :)